### ISSUE TITLE
add a linter to avoid asserting whether slice is empty (or not) by `arr == nil`.

### ISSUE THREAD

**Issue by YangKeao** [pingcap/tidb#52074]
## Enhancement

Ref #51807. It's easy to make mistake that `[]T{} == nil` returns false, so that we'd better to always use `len(arr) > 0` to assert whether a slice is empty. We should also provide an utility to assert whether the slice is really nil.

**Comment by lance6716** [comment]
it's oversimplify to check `== nil` to avoid #51807. One point is even if we have `emptynil.IsNilMap` we may still write `if !emptynil.IsNilMap(nt.ForeignKeys)` and the bug will still happen. Another point is checking the zero value (nil) is common in use cases as variable lazily initialization, iterator reaching end, and so on. It's not worth to bother changing these more widely used cases.

For #51807 I can't find a systematic way to prevent it. Only the developer knows if the state of zero length container will occur and should occur.

**Comment by YangKeao** [comment]
> it's oversimplify to check `== nil` to avoid #51807. One point is even if we have `emptynil.IsNilMap` we may still write `if !emptynil.IsNilMap(nt.ForeignKeys)` and the bug will still happen. Another point is checking the zero value (nil) is common in use cases as variable lazily initialization, iterator reaching end, and so on. It's not worth to bother changing these more widely used cases.
> 
> For #51807 I can't find a systematic way to prevent it. Only the developer knows if the state of zero length container will occur and should occur.

@lance6716  I totally agree with you, and I trust the developers 100%. There is no way to tell the original expectation of the developer (unless we analyze the codes manually one by one for these 1.5k cases, which is impossible and not valuable.)

The problem is that `== nil` and `len(xx) == 0` have different semantic, but they are easy to be mis-used. This linter is here to ask the developers to **think twice** before writing `== nil` and I believe, after thinking about this issue, the developers can write a correct behavior. Even if we don't, it's easier to debug in the future as we distinguish these two different semantics.

**Comment by lance6716** [comment]
I guess https://github.com/pingcap/tidb/issues/51807 is caused by `delete(mapVar, key)`? Do you think checking the map `delete` and resetting a zero entry map to `nil` will cause less burden?

**Comment by D3Hunter** [comment]
i think it's a little over-reacting to the bug in description. misusing between `len(xx) == 0` and `xx == nil` are human errors, that cannot be avoided by replacing `xx == nil` with `isNil(xx)`, we still might make the same mistake by misusing `len(xx) == 0` and `isNil(xx)`, and it sacrifices readability.

https://github.com/pingcap/tidb/pull/52073#issuecomment-2021997624

**Comment by wuhuizuo** [comment]
Some ref links i searched:
- https://stackoverflow.com/questions/38144050/how-do-i-check-for-an-empty-slice
- https://www.includehelp.com/golang/what-is-the-easiest-way-to-check-if-a-slice-is-empty-in-golang.aspx

Any cases which prefer `xx == nil` rather than `len(xx) == 0` to judge a slice?
